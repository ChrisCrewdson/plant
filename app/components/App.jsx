const React = require('react');
const actions = require('../actions');
const PropTypes = require('prop-types');

class App extends React.Component {
  static contextTypes = {
    store: PropTypes.object.isRequired,
  };

  componentWillMount() {
    const { store } = this.context;
    const users = store.getState().get('users');
    if (!users || users.size === 0) {
      store.dispatch(actions.loadUsersRequest());
    }

    const locations = store.getState().get('locations');
    if (!locations || locations.size === 0) {
      store.dispatch(actions.loadLocationsRequest());
    }

    // const user = store.getState().get('user');
    // if(user && user.get('isLoggedIn', false)) {
    //   const locationIds = user.get('locationIds');
    //   if(!locationIds) {
    //     const userId = user.get('_id');
    //     store.dispatch(actions.loadLocationsRequest(userId));
    //   }
    // }
  }

  render() {
    return (
      <div className="react-root">
        {this.props.children}
      </div>
    );
  }
}

App.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  children: PropTypes.object.isRequired,
};

module.exports = App;
