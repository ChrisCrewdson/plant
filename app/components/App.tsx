import React from 'react';
import PropTypes from 'prop-types';
import { actionFunc } from '../actions';

export default class App extends React.Component {
  // TODO: When tsc 3.7+ is in use remove the ! to see hint text on how to change this.
  context!: PlantContext;

  static propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    children: PropTypes.object.isRequired,
  };

  static contextTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    store: PropTypes.object.isRequired,
  };

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    const { store } = this.context;
    const { users = {}, locations = {} } = store.getState();

    // TODO: This will cause a problem for a non-initialized site
    // that has zero users and zero locations as these values
    // will be 0 which is falsy.
    const usersCount = Object.keys(users).length;
    const locationsCount = Object.keys(locations).length;

    if (!usersCount) {
      store.dispatch(actionFunc.loadUsersRequest());
    }

    if (!locationsCount) {
      store.dispatch(actionFunc.loadLocationsRequest());
    }
  }

  render() {
    const { children } = this.props;
    return (
      <div className="react-root">
        {children}
      </div>
    );
  }
}
