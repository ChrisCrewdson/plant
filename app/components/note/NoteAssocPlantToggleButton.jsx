// Wrapper for the Raised Button to make it toggle for the Plants
// that are associated with a note.
// TODO: Make this a generic ToggleRaisedButton component
//       and move it to the common folder.
const React = require('react');
const RaisedButton = require('material-ui/RaisedButton').default;
const PropTypes = require('prop-types');

class NoteAssocPlantToggleButton extends React.PureComponent {
  constructor() {
    super();
    this.toggle = this.toggle.bind(this);
  }

  toggle() {
    const { toggleFunc, _id } = this.props;
    toggleFunc(_id);
  }

  render() {
    const {
      label,
      primary,
      secondary,
      style,
    } = this.props;

    return (
      <RaisedButton
        label={label}
        onClick={this.toggle}
        primary={primary}
        secondary={secondary}
        style={style}
      />
    );
  }
}

NoteAssocPlantToggleButton.propTypes = {
  _id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  primary: PropTypes.bool.isRequired,
  secondary: PropTypes.bool.isRequired,
  style: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  toggleFunc: PropTypes.func.isRequired,
};

module.exports = NoteAssocPlantToggleButton;
