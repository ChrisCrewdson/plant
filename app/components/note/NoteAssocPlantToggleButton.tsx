// Wrapper for the Raised Button to make it toggle for the Plants
// that are associated with a note.
// TODO: Make this a generic ToggleRaisedButton component
//       and move it to the common folder.
import React from 'react';
import PropTypes from 'prop-types';

import Button from '@material-ui/core/Button';
import MaterialCore from '@material-ui/core';

type NoteAssocPlantToggleButtonType =
  'selected' |
  'alive' |
  'dead';

interface NoteAssocPlantToggleButtonProps {
  _id: string;
  label: React.ReactNode;
  selectState: NoteAssocPlantToggleButtonType;
  style: React.CSSProperties;
  toggleFunc: (id: string) => void;
}

export default class NoteAssocPlantToggleButton extends React.PureComponent {
  props!: NoteAssocPlantToggleButtonProps;

  /**
   * The 3 colors that I settled on for associated a note with plants are:
   * Selected - primary color - blue - this means that the plant is tagged on this note
   * Alive - default color - transparent/white - the plant is not tagged and is alive
   * Dead - secondary color - red/pink/orange - the plant is dead and not tagged.
   */
  colors: Record<NoteAssocPlantToggleButtonType, MaterialCore.PropTypes.Color> = {
    alive: 'default',
    dead: 'secondary',
    selected: 'primary',
  };

  static propTypes = {
    _id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selectState: PropTypes.string.isRequired,
    style: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
    toggleFunc: PropTypes.func.isRequired,
  };

  constructor(props: NoteAssocPlantToggleButtonProps) {
    super(props);
    this.toggle = this.toggle.bind(this);
  }

  toggle() {
    const { toggleFunc, _id } = this.props;
    toggleFunc(_id);
  }

  render() {
    const {
      label,
      selectState,
      style,
    } = this.props;
    const color = this.colors[selectState];

    const buttonStyle = { ...style, fontSize: 'medium' };

    return (
      <Button
        onClick={this.toggle}
        color={color}
        size="large"
        style={buttonStyle}
        variant="contained"
      >
        {label}
      </Button>
    );
  }
}
