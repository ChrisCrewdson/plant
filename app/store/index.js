const { createStore, applyMiddleware } = require('redux');
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');
const Immutable = require('immutable');

let middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  const logger = require('../middleware/logger');
  middleware = [logger, api];
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

const store = createStoreWithMiddleware(reducers, Immutable.fromJS(window.__INITIAL_STATE__ || {}));

userSubscribe(store);

module.exports = store;
