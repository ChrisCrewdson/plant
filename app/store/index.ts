export {}; // To get around: Cannot redeclare block-scoped variable .ts(2451)

const { createStore, applyMiddleware } = require('redux');

const seamless = require('seamless-immutable').static;
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');

const middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  const logger = require('../middleware/logger');
  middleware.unshift(logger);
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

// @ts-ignore - __INITIAL_STATE__ is added on the server
const initialState = seamless.from(window.__INITIAL_STATE__ || {});
const store = createStoreWithMiddleware(reducers, initialState);

userSubscribe(store);

module.exports = store;
