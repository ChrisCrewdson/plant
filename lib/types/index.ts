
// Definitions file for Plant Project

// Currently NonEmptyList is not used anywhere.
// Plan to tag the plantIds in a Note with this as there
// always has to be at least 1 plant for each note.
type NonEmptyList<T> = T[] & { 0: T };

// Now included with Typescript 3.5+
// type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

/*
In TypeScript, just as in ECMAScript 2015, any file containing a top-level import or export
is considered a module. Conversely, a file without any top-level import or export declarations
is treated as a script whose contents are available in the global scope (and therefore to
modules as well).
*/

type GeoType = 'Point';

interface GeoCoords {
  /** Longitude */
  0: number;
  /** Latitude */
  1: number;
}

interface Geo {
  type: GeoType; // "Point" - is this the only value I know of that's valid
  coordinates: GeoCoords;
}
interface Window {
  [key: string]: any;
  // FormData: Function;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

interface Global extends NodeJS.Global {
  [key: string]: any;
  // window: Window;
  // requestAnimationFrame?: (callback: Function) => void;
}

type Html5InputTypes =
  | 'button'
  | 'checkbox'
  | 'color'
  | 'date'
  | 'datetime-local'
  | 'email'
  | 'file'
  | 'hidden'
  | 'image'
  | 'month'
  | 'number'
  | 'password'
  | 'radio'
  | 'range'
  | 'reset'
  | 'search'
  | 'submit'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week';
