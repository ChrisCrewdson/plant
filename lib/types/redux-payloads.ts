// PlantRedux
// This file holds the specialized typings that apply to this project for Redux operations.

import { Action } from 'redux';

/**
 * The Actions in Plant have 2 props:
 * type and payload
 * type is always a string so we can extend redux.Action for this.
 * payload can be anything
 */
export interface PlantAction<T = any> extends Action<string> {
    payload: T;
  }

export interface LoadLocationsSuccessAction extends PlantAction<UiLocationsValue[]> {}
