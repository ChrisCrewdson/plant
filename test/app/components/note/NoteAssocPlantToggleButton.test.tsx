import React from 'react';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { Dispatch } from 'redux';
import NoteAssocPlantToggleButton from '../../../../app/components/note/NoteAssocPlantToggleButton';
import store from '../../../../app/store';
import App from '../../../../app/components/App';

import { PlantAction } from '../../../../lib/types/redux-payloads';

const muiTheme = getMuiTheme(lightBaseTheme);

describe('NoteAssocPlantToggleButton', () => {
  const storeDispatch = store.dispatch;

  beforeAll(() => {
    store.dispatch = (() => {}) as Dispatch<PlantAction<any>>;
  });

  afterAll(() => {
    store.dispatch = storeDispatch;
  });

  test('NoteAssocPlantToggleButton should be rendered', () => {
    const component = renderer.create(
      <MuiThemeProvider muiTheme={muiTheme}>
        <Provider store={store}>
          <App>
            <MemoryRouter>
              <NoteAssocPlantToggleButton
                _id="buttonId"
                label="Button Label"
                primary
                secondary={false}
                style={{ color: 'blue' }}
                toggleFunc={() => {}}
              />
            </MemoryRouter>
          </App>
        </Provider>
      </MuiThemeProvider>);
    const badgeTotals = component.toJSON();
    expect(badgeTotals).toMatchSnapshot();
  });
});
