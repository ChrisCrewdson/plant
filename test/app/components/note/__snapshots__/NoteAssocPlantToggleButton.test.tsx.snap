// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteAssocPlantToggleButton NoteAssocPlantToggleButton should be rendered 1`] = `
<div
  className="react-root"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "color": "blue",
        "fontSize": "medium",
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Button Label
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;
